### PR TITLE
[pbh]: Implement Edit Flows

### PR DESCRIPTION
**Pull request by nazariig**
Signed-off-by: Nazarii Hnydyn <nazariig@nvidia.com>

PBH Edit Flows is a second phase of PBH feature implementation. It allows user to modify the already existing objects.
PBH Edit Flows offer a full entity update which assumes Config DB field ADD/UPDATE/REMOVE processing. PBH OA interprets series of Redis HSET/HDEL as SONiC SET operation and internally does the diff of existing/supplied configuration to deploy the new changes.

HLD: https://github.com/Azure/SONiC/pull/909

**What I did**
* Implemented Edit Flows in scope of PBH enhancement

**Why I did it**
* Implementation is done according to the PBH HLD

**How I verified it**
* N/A

**Details if related**
* N/A

### CODE REVIEW

**Comment by nazariig** [comment]
@anish-n / @kktheballer FYI

**Comment by liat-grozovik** [comment]
@anish-n  please review and provide your feedback.

**Comment by nazariig** [comment]
@anish-n please have a look

**Comment by nazariig** [comment]
@anish-n just a kind reminder

**Review by stepanblyschak — DISMISSED** [review]

**Comment by nazariig** [comment]
@anish-n just a kind reminder

File: orchagent/pbh/pbhcnt.h [hunk 1]
```diff
@@ -105,19 +105,22 @@ class PbhRule final : public PbhContainer
...
+            std::string name;
+        } meta;
         std::string value;
         bool is_set = false;
     } hash;
 
     struct {
+        struct {
+            std::string name;
+        } meta;
```
[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcnt.h:118]: @anish-n has been already discussed

File: orchagent/pbh/pbhcap.cpp [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+    this->setPbhDefaults(this->rule.l4_dst_port);
+    this->setPbhDefaults(this->rule.inner_ether_type);
+    this->rule.hash.insert(PbhFieldCapability::UPDATE);
+    this->setPbhDefaults(this->rule.packet_action);
+    this->setPbhDefaults(this->rule.flow_counter);
+
+    this->hash.hash_field_list.insert(PbhFieldCapability::UPDATE);
+}
+
+PbhMellanoxFieldCapabilities::PbhMellanoxFieldCapabilities() noexcept
```
[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcap.cpp:126]: @anish-n has been already discussed

File: orchagent/pbh/pbhcap.cpp [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+static const std::map<PbhFieldCapability, std::string> pbhFieldCapabilityMap =
+{
+    { PbhFieldCapability::ADD,    PBH_FIELD_CAPABILITY_ADD },
+    { PbhFieldCapability::UPDATE, PBH_FIELD_CAPABILITY_UPDATE },
+    { PbhFieldCapability::REMOVE, PBH_FIELD_CAPABILITY_REMOVE }
+};
+
+// functions ----------------------------------------------------------------------------------------------------------
+
+static inline std::string toStr(PbhAsicVendor value) noexcept
```
[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcap.cpp:55]: @anish-n the main purpose of inline here is to notify user that code is pretty simple to use it as function. If you believe that it is better to leave as regular one - please let me know

File: orchagent/pbh/pbhcap.cpp [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+    const auto &cit = pbhAsicVendorMap.find(value);
+    if (cit != pbhAsicVendorMap.cend())
+    {
+        return cit->second;
+    }
+
+    return PBH_PLATFORM_UNKN;
+}
+
+static inline std::string toStr(PbhFieldCapability value) noexcept
```
[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcap.cpp:66]: @anish-n the main purpose of inline here is to notify user that code is pretty simple to use it as function. If you believe that it is better to leave as regular one - please let me know

File: orchagent/pbhorch.cpp [hunk 5]
```diff
@@ -577,57 +665,227 @@ bool PbhOrch::updatePbhRule(const PbhRule &rule)
...
+        {
+            auto pbhRulePtr = dynamic_cast<AclRulePbh*>(this->aclOrch->getAclRule(rule.table, rule.name));
+
+            if (pbhRulePtr == nullptr)
+            {
+                SWSS_LOG_ERROR("Failed to update PBH rule(%s) in SAI: invalid object type", rule.key.c_str());
+                return false;
+            }
+
+            if (!pbhRulePtr->disableAction())
```
Comment (anish-n) [inline orchagent/pbhorch.cpp:857]: So if I understand the work around correctly, we will disable the acl  rule and then update the acl rule, which will as part of the update re-enable the acl with the newly defined rule?

File: orchagent/pbhorch.cpp [hunk 6]
```diff
@@ -252,6 +312,34 @@ bool PbhOrch::updatePbhTable(const PbhTable &table)
...
+    if (!this->pbhCap.validatePbhTableCap(aFields, PbhFieldCapability::ADD))
+    {
+        SWSS_LOG_ERROR("Failed to validate PBH table(%s) added fields: unsupported capabilities", table.key.c_str());
+        return false;
+    }
+
+    if (!this->pbhCap.validatePbhTableCap(uFields, PbhFieldCapability::UPDATE))
+    {
+        SWSS_LOG_ERROR("Failed to validate PBH table(%s) updated fields: unsupported capabilities", table.key.c_str());
+        return false;
```
Comment (anish-n) [inline orchagent/pbhorch.cpp:334]: Does return false mean that the operation will keep retrying in orchagent?

**Review by anish-n — DISMISSED** [review]
Overall looks good to me, just have a few questions which I would like to understand on

File: orchagent/pbhorch.cpp (hunk 5 above)
Comment (nazariig) [reply orchagent/pbhorch.cpp:857]: PBH rule configuration defines `packet_action` and `hash` attributes as:
1. SAI_ACL_ENTRY_ATTR_ACTION_SET_LAG_HASH_ID
2. SAI_ACL_ENTRY_ATTR_ACTION_SET_ECMP_HASH_ID

Which means that two PBH fields refer to a single separate action from SAI point of view.

**Example:**
```json
"PBH_RULE": {
    "pbh_table|nvgre": {
        "priority": "2",
            "ether_type": "0x0800",
            "ip_protocol": "0x2f",
            "gre_key": "0x2500/0xffffff00",
            "inner_ether_type": "0x86dd",
            "hash": "inner_v6_hash",
            "packet_action": "SET_ECMP_HASH",
            "flow_counter": "DISABLED"
    }
}
```

In order to implement a W/A for Mellanox platform, we need to disable both actions first, before moving to update (`packet_action` or `hash`) operation via ACL orch:

```cpp
bool AclRulePbh::disableAction()
{
    const auto &cit1 = m_actions.find(SAI_ACL_ENTRY_ATTR_ACTION_SET_ECMP_HASH_ID);
    if (cit1 != m_actions.cend())
    {
        const auto &attr1 = cit1->second.getSaiAttr();
        if (attr1.value.aclaction.enable)
        {
            sai_attribute_t attr;

            attr.id = attr1.id;
            attr.value.aclaction.enable = false;
            attr.value.aclaction.parameter.oid = SAI_NULL_OBJECT_ID;

            if (!setAttribute(attr))
            {
                return false;
            }

            m_actions.erase(cit1);
        }
    }

    const auto &cit2 = m_actions.find(SAI_ACL_ENTRY_ATTR_ACTION_SET_LAG_HASH_ID);
    if (cit2 != m_actions.cend())
    {
        const auto &attr2 = cit2->second.getSaiAttr();
        if (attr2.value.aclaction.enable)
        {
            sai_attribute_t attr;

            attr.id = attr2.id;
            attr.value.aclaction.enable = false;
            attr.value.aclaction.parameter.oid = SAI_NULL_OBJECT_ID;

            if (!setAttribute(attr))
            {
                return false;
            }

            m_actions.erase(cit2);
        }
    }

    return true;
}
```

File: orchagent/pbh/pbhcap.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (anish-n) [reply orchagent/pbh/pbhcap.cpp:55]: This should be a regular function, don't believe inline is well suited, same comment for other uses of inline in this pr, typically I have seen inline defined .h files and for simpler functions.

[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcap.cpp]: @anish-n done

File: orchagent/pbh/pbhcap.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (nazariig) [reply orchagent/pbh/pbhcap.cpp]: @anish-n done

**Comment by nazariig** [comment]
@qiluo-msft / @prsunny the review is done. Please help to merge

**Comment by prsunny** [comment]
@anish-n , can you please sign-off?

**Review by anish-n — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@nazariig  can you please extend your PR description and also refer to the HLD itself?

**Comment by nazariig** [comment]
> @nazariig can you please extend your PR description and also refer to the HLD itself?

@liat-grozovik the link to HLD is already there and the description has been already updated

**Review by stepanblyschak — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Pull request merged**